File: common/src/main/java/net/typho/vibrancy/BlockStateFunction.java
package net.typho.vibrancy;

import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import net.minecraft.world.level.block.Block;
import net.minecraft.world.level.block.state.BlockState;
import org.apache.commons.lang3.tuple.ImmutablePair;

import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.function.Function;
import java.util.function.Predicate;
import java.util.function.Supplier;

public interface BlockStateFunction<T> {
    T apply(BlockState state);

    static <T> BlockStateFunction<T> parse(Map<Predicate<BlockState>, Supplier<T>> map, Supplier<T> def) {
        return state -> map.entrySet().stream()
                .filter(entry -> entry.getKey().test(state))
                .map(Map.Entry::getValue)
                .findAny()
                .orElse(def)
                .get();
    }

    static <T> BlockStateFunction<T> parseJson(Block block, JsonElement json, Function<JsonElement, T> toT, Supplier<T> def) {
        if (json.isJsonObject()) {
            JsonObject obj = json.getAsJsonObject();
            JsonElement defElement = obj.get("default");

            if (defElement != null) {
                T t = toT.apply(defElement);
                def = () -> t;
            }

            return parse(
                    obj.asMap().entrySet().stream()
                            .filter(entry -> !entry.getKey().equals("default"))
                            .map(entry -> {
                                T value = toT.apply(entry.getValue());
                                return new ImmutablePair<Predicate<BlockState>, Supplier<T>>(Vibrancy.BLOCK_STATE_PREDICATE.apply(block.getStateDefinition(), entry.getKey()), () -> value);
                            })
                            .collect(
                                    LinkedHashMap::new,
                                    (map, value) -> map.put(value.getLeft(), value.getRight()),
                                    HashMap::putAll
                            ),
                    def
            );
        } else {
            T t = toT.apply(json);
            return state -> t;
        }
    }
}

File: common/src/main/kotlin/net/typho/vibrancy/api/BlockStateFunction.kt
package net.typho.vibrancy.api

import net.minecraft.world.level.block.state.BlockState

data class BlockStateFunction<T>(
    val entries: List<Entry<T>>,
    val default: T
) {
    constructor(default: T, vararg entries: Entry<T>) : this(entries.toList(), default)

    fun apply(state: BlockState): T {
        for (entry in entries) {
            if (entry.test(state)) {
                return entry.value
            }
        }

        return default
    }

    data class Entry<T>(
        val map: Map<String, Comparable<*>>,
        val value: T
    ) {
        constructor(property: String, comparable: Comparable<*>, value: T) : this(mapOf(Pair(property, comparable)), value)

        fun test(state: BlockState): Boolean {
            for (entry in map.entries) {
                if (state.getValue(state.block.stateDefinition.getProperty(entry.key)) != entry.value) {
                    return false
                }
            }

            return true
        }
    }
}